crates/primitives/src/forkid.rs

```diff
@@ -39,14 +41,18 @@ impl fmt::Debug for ForkHash {
 
 impl From<H256> for ForkHash {
     fn from(genesis: H256) -> Self {
-        Self(crc32::checksum_ieee(&genesis[..]).to_be_bytes())
+        Self(CRC_32_IEEE.checksum(&genesis[..]).to_be_bytes())
     }
 }
 
 impl AddAssign<BlockNumber> for ForkHash {
     fn add_assign(&mut self, block: BlockNumber) {
         let blob = block.to_be_bytes();
-        self.0 = crc32::update(u32::from_be_bytes(self.0), &crc32::IEEE_TABLE, &blob).to_be_bytes();
+        #[allow(clippy::suspicious-op-assign-impl)]
+        let initial = (u32::from_be_bytes(self.0) ^ CRC_32_IEEE.algorithm.xorout).reverse_bits();
```

Comment:
this is new?

why?

Reply:
Because cargo clippy in github workflow is failing, so  I had to add this

Reply:
I was referring to the additional XOR step, seems like a weird API change

Reply:
Yes, crc 3.x has a lot of breaking API changes

Reply:
Is it OK to resolve this conversation now?

Reply:
This does feel weird https://docs.rs/crc/latest/crc/struct.Algorithm.html#structfield.xorout.

Shouldn't we instead be doing `self.digest()` instead of `_with_initial` and then calling `.update(self.0)`?

Reply:
Do you mean the following logic:

```rust
let mut digest = CRC_32_IEEE.digest();
// self.0.reverse();
digest.update(&self.0);
```

It can NOT pass tests.

Reply:
even though I don't understand how this crc works and find the API super weird, it still passes the tests.
but I think it's very odd that we now have to do XOR and reverse bits manually when it previously was just an update call.

Reply:
Yes, the APIs are very obscure, took me days to pass all tests

Reply:
The problem here is the `Digest` API doesn't quite fit for our use case because it forces us to keep finalizing and initializing, rather than obtaining the value produced when `update` is run (this is the value we care about).

The line with the XOR is effectively un-finalizing the CRC, and reversing the bits before they are reversed when the CRC is initialized. After this is done, the CRC is in the same state it would have been in, had we initialized the CRC and called `update` on previous values. For a given `value` in this CRC, finalization is basically this:
```rust
xorout ^ value
```
and initialization is this:
```rust
value.reverse_bits()
```

Maybe we could do the following, because initialization and finalization cancel each other out:
```rust
        let digest = CRC_32_IEEE.digest_with_initial(u32::from_be_bytes(self.0));
        let value = digest.finalize();
        let mut digest = CRC_32_IEEE.digest_with_initial(value);
        digest.update(&blob);
```

Reply:
Thanks @Rjected , your version all works and looks much cleaner, I've updated my code to your version.